src/adaptivity.jl

```diff
@@ -1,85 +1,207 @@
+end
+
+# S forward Interpolation
+function S_interpolate(t, coeffs)
+	ts = [t^(i - 1) for i in axes(coeffs, 2)]
+	return coeffs * ts
+end
+
+function dS_interpolate(t, S_coeffs)
+	ts = zeros(size(S_coeffs, 2))
+	for i in 2:size(S_coeffs, 2)
+		ts[i] = (i - 1) * t^(i - 2)
+	end
+	return S_coeffs * ts
 end
 
 """
-    interval(mesh, t)
+	interval(mesh, t)
 
 Find the interval that `t` belongs to in `mesh`. Assumes that `mesh` is sorted.
 """
 function interval(mesh, t)
-    return clamp(searchsortedfirst(mesh, t) - 1, 1, length(mesh) - 1)
+	return clamp(searchsortedfirst(mesh, t) - 1, 1, length(mesh) - 1)
 end
 
 """
-    mesh_selector!(cache::MIRKCache)
+	mesh_selector!(cache::MIRKCache)
 
 Generate new mesh based on the defect.
 """
-@views function mesh_selector!(cache::MIRKCache{iip, T}) where {iip, T}
+@views function mesh_selector!(cache::Union{MIRKCache{iip, T}, FIRKCacheExpand{iip, T}, FIRKCacheNested{iip, T}}) where {iip, T}
     (; M, order, defect, mesh, mesh_dt) = cache
     (_, MxNsub, abstol, _, _), kwargs = __split_mirk_kwargs(; cache.kwargs...)
     N = length(cache.mesh)
 
-    safety_factor = T(1.3)
-    ρ = T(1.0) # Set rho=1 means mesh distribution will take place everytime.
-    Nsub_star = 0
-    Nsub_star_ub = 4 * (N - 1)
-    Nsub_star_lb = N ÷ 2
-
-    info = ReturnCode.Success
-
-    ŝ = [maximum(abs, d) for d in defect]  # Broadcasting breaks GPU Compilation
-    ŝ .= (ŝ ./ abstol) .^ (T(1) / (order + 1))
-    r₁ = maximum(ŝ)
-    r₂ = sum(ŝ)
-    r₃ = r₂ / (N - 1)
-
-    n_predict = round(Int, (safety_factor * r₂) + 1)
-    n = N - 1
-    n_ = T(0.1) * n
-    n_predict = ifelse(abs((n_predict - n)) < n_, round(Int, n + n_), n_predict)
-
-    if r₁ ≤ ρ * r₂
-        Nsub_star = 2 * (N - 1)
-        if Nsub_star > MxNsub # Need to determine the too large threshold
-            info = ReturnCode.Failure
-            meshₒ = mesh
-            mesh_dt₀ = mesh_dt
-        else
-            meshₒ = copy(mesh)
-            mesh_dt₀ = copy(mesh_dt)
-            half_mesh!(cache)
-        end
-    else
-        Nsub_star = clamp(n_predict, Nsub_star_lb, Nsub_star_ub)
-        if Nsub_star > MxNsub
-            # Mesh redistribution fails
-            info = ReturnCode.Failure
-            meshₒ = mesh
-            mesh_dt₀ = mesh_dt
-        else
-            ŝ ./= mesh_dt
-            meshₒ = copy(mesh)
-            mesh_dt₀ = copy(mesh_dt)
-            redistribute!(cache, Nsub_star, ŝ, meshₒ, mesh_dt₀)
-        end
-    end
-    return meshₒ, mesh_dt₀, Nsub_star, info
+	safety_factor = T(1.3)
+	ρ = T(1.0) # Set rho=1 means mesh distribution will take place everytime.
+	Nsub_star = 0
+	Nsub_star_ub = 4 * (N - 1)
+	Nsub_star_lb = N ÷ 2
+
+	info = ReturnCode.Success
+
+	ŝ = [maximum(abs, d) for d in defect]  # Broadcasting breaks GPU Compilation
+	ŝ .= (ŝ ./ abstol) .^ (T(1) / (order + 1))
+	r₁ = maximum(ŝ)
+	r₂ = sum(ŝ)
+	r₃ = r₂ / (N - 1)
+
+	n_predict = round(Int, (safety_factor * r₂) + 1)
+	n = N - 1
```

Comment:
I will review more carefully later but Tabs need to be changed to spaces

Reply:
Thanks! I changed this with the latest commit